homekit/controller/tools.py

```diff
@@ -166,3 +173,138 @@ def check_convert_value(val, target_type):
+        Create a DelayedExecution for the given function. The function will be executed after the given timeout (in
+        seconds, fractions of seconds are also possible).
+        :param func: the function to call
+        :param timeout: the time out in seconds
+        """
+        threading.Thread.__init__(self)
+        self.timeout = timeout
+        self.function = func
+
+    def run(self):
+        time.sleep(self.timeout)
+        self.function()
+
+
+def parse_manufacturer_specific_data(input_data):
+    """
+    Parse the manufacturer specific data as returned via Bluez ManufacturerData. This skips the data for LEN, ADT and
+    CoID as specified in Chapter 6.4.2.2 of the spec on page 124. Data therefore starts at TY (must be 0x06).
+
+    :param input_data: manufacturer specific data as bytes
+    :return: a dict containing the type (key 'type', value 'HomeKit'), the status flag (key 'sf'), human readable
+             version of the status flag (key 'flags'), the device id (key 'device_id'), the accessory category
+             identifier (key 'acid'), human readable version of the category (key 'category'), the global state number
+             (key 'gsn'), the configuration number (key 'cn') and the compatible version (key 'cv')
+    """
+    logging.debug('manufacturer specific data: %s', input_data.hex())
+
+    # the type must be 0x06 as defined on page 124 table 6-29
+    ty = input_data[0]
+    input_data = input_data[1:]
+    if ty == 0x06:
+        ty = 'HomeKit'
+
+        ail = input_data[0]
+        logging.debug('advertising interval %s', '{0:02x}'.format(ail))
+        length = ail & 0b00011111
+        if length != 13:
+            logging.debug('error with length of manufacturer data')
+        input_data = input_data[1:]
+
+        sf = input_data[0]
+        if sf == 0:
+            flags = 'paired'
+        elif sf == 1:
+            flags = 'unpaired'
+        else:
+            flags = 'error'
+        input_data = input_data[1:]
+
+        device_id = (':'.join(input_data[:6].hex()[0 + i:2 + i] for i in range(0, 12, 2))).upper()
+        input_data = input_data[6:]
+
+        acid = int.from_bytes(input_data[:2], byteorder='little')
+        input_data = input_data[2:]
+
+        gsn = int.from_bytes(input_data[:2], byteorder='little')
+        input_data = input_data[2:]
+
+        cn = input_data[0]
+        input_data = input_data[1:]
+
+        cv = input_data[0]
+        input_data = input_data[1:]
+        if len(input_data) > 0:
+            logging.debug('remaining data: %s', input_data.hex())
+        return {'type': ty, 'sf': sf, 'flags': flags, 'device_id': device_id, 'acid': acid, 'gsn': gsn, 'cn': cn,
+                'cv': cv, 'category': Categories[int(acid)]}
+
+    return {'manufacturer': 'apple', 'type': ty}
+
+
+# 0x004c is the Company Identifier code for Apple Inc. (see Chapter 6.4.2.2 of the spec on page 124)
+COID_APPLE = 0x004c
+
+
+class HomekitDiscoveryDevice(gatt.Device):
```

Comment:
Note that in my branch this is implemented on my main Device subclass. This is because I've found that the discovery data is useful for more than just discovery, and I want to be able to access gsn/cn fields from the events code in particular.